### PR TITLE
Add PHPStan\dumpPhpDocType()

### PR DESCRIPTION
**Pull request by zonuexe**
resolve https://github.com/phpstan/phpstan/issues/11823, refs https://github.com/phpstan/phpstan-src/pull/3555

### CODE REVIEW

File: src/Rules/Debug/DumpPhpDocTypeRule.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+		if (count($node->getArgs()) === 0) {
+			return [];
+		}
+
+		return [
+			RuleErrorBuilder::message(
+				sprintf(
+					'Dumped type: %s',
+					$scope->getType($node->getArgs()[0]->value)->toPhpDocNode(),
```
Comment (ondrejmirtes) [inline src/Rules/Debug/DumpPhpDocTypeRule.php]: Use the Printer here (https://phpstan.github.io/phpdoc-parser/2.0.x/PHPStan.PhpDocParser.Printer.Printer.html#_print), not `Node::__toString()`. It's more precise, uses less parentheses, only where needed.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Rules/Debug/DumpPhpDocTypeRuleTest.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+	public function testRuleSymbols(): void
+	{
+		$this->analyse([__DIR__ . '/data/dump-phpdoc-type.php'], [
+			[
+				"Dumped type: array{'': ''}",
+				5,
+			],
+			[
+				"Dumped type: array{'\0': 'NUL', NUL: '\0'}",
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Debug/DumpPhpDocTypeRuleTest.php:28]: Can you also try some of the special multi-byte characters you did in the previous PR? (`\x??`)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/dumpType.php [hunk 3]
```diff
@@ -13,3 +13,15 @@ function dumpType($value) // phpcs:ignore Squiz.Functions.GlobalFunction.Found
...
+/**
+ * @phpstan-pure
+ * @param mixed $value
+ * @return mixed
+ *
+ * @throws void
+ */
+function dumpPhpDocType($value) // phpcs:ignore Squiz.Functions.GlobalFunction.Found
+{
+	return null;
```
Comment (ruudk) [inline src/dumpType.php:26]: I know this is copied from `dumpType`, but I always wondered: Why does this even return something? 

PHPStorm always complains that the result is not used.

Can't we make this return `void`? Then do the same for the other helpers?

/cc @ondrejmirtes

Comment (ondrejmirtes) [reply src/dumpType.php:26]: We can’t make void functions pure, that’s why.

Personally I turn this off in PhpStorm, as PHPStan will tell me the same thing.

Comment (zonuexe) [reply src/dumpType.php:26]: I think this is because if declared as void, `dumpType()` would affect the purity of the embedded function.

**Comment by ondrejmirtes** [comment]
Please fix CS https://github.com/phpstan/phpstan-src/actions/runs/11252140969/job/31284635295#step:7:1

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by staabm** [comment]
@zonuexe please see PHP7.4 only test errors

grafik


https://github.com/phpstan/phpstan-src/actions/runs/11254262628/job/31291308442?pr=3561
https://github.com/phpstan/phpstan-src/actions/runs/11254262628/job/31291309269?pr=3561

interesstingly these appear in [a followup PR](https://github.com/phpstan/phpstan-src/pull/3561), and were not detected in this PR here.

maybe its a PHPStan 2.x only thing

**Pull request merged**